### PR TITLE
Generic Ubuntu VM script

### PR DESCRIPTION
**Pull request by irish1986**
> [!NOTE]
> We are meticulous when it comes to merging code into the main branch, so please understand that we may reject pull requests that do not meet the project's standards. It's never personal. Also, game-related scripts have a lower chance of being merged.

## Description

Create a new generic ubuntu-vm script that allows the user ot select which version of ubuntu should be installed.

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [x] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [X] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [X] Testing performed (I have tested my changes, ensuring everything works as expected)
- [X] Documentation updated (I have updated any relevant documentation)

## Additional Information (optional)
C&C are welcomed via PR comments.

## Related Pull Requests / Discussions

If there are other pull requests or discussions related to this change, please link them here:
- Related Discussion #268

### CODE REVIEW

**Comment by Mellowlynx** [comment]
I don't see the need for this script, To choose a version I have to go to all the advanced options, there are separate scripts for this reason.
Currently, you run the script of the version you want, not extra options needed.
If the update was just a version question in the standard too, maybe.

**Comment by havardthom** [comment]
Ubuntu focal has end of standard support april 2025, we do not want to add that option now. Oracular is not LTS. 

I'm not totally opposed to having one Ubuntu VM script which defaults to noble and with the option to downgrade to jammy in advanced options. This would make it consistent with the Ubuntu LXC (where we should update default version to noble). 

This is also a VM script so it should not be a breaking change to existing users.

**Comment by MickLesk** [comment]
>
> This is also a VM script so it should not be a breaking change to existing users.
> 
> 



What is your opinion? I've skimmed through it and am rather against a merge. That's what the advanced settings are for.

File: vm/ubuntu-vm.sh [hunk 1]
```diff
@@ -2,17 +2,18 @@
 
 # Copyright (c) 2021-2024 tteck
 # Author: tteck (tteckster)
+#         irish1986 (irish1986)
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:5]:
```suggestion
# Co-Author: irish1986 (irish1986)
```

File: vm/ubuntu-vm.sh [hunk 2]
```diff
@@ -21,6 +22,11 @@ echo -e "\n Loading..."
 GEN_MAC=02:$(openssl rand -hex 5 | awk '{print toupper($0)}' | sed 's/\(..\)/\1:/g; s/.$//')
 NEXTID=$(pvesh get /cluster/nextid)
 
+ORACULAR="oracular"
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:25]:
```suggestion
```

File: vm/ubuntu-vm.sh [hunk 3]
```diff
@@ -21,6 +22,11 @@ echo -e "\n Loading..."
...
+NOBLE="noble"
+JAMMY="jammy"
+FOCAL="focal"
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:28]:
```suggestion
```

File: vm/ubuntu-vm.sh [hunk 4]
```diff
@@ -155,10 +162,21 @@ function default_settings() {
   echo -e "${DGN}Using VLAN: ${BGN}Default${CL}"
   echo -e "${DGN}Using Interface MTU Size: ${BGN}Default${CL}"
   echo -e "${DGN}Start VM when completed: ${BGN}no${CL}"
-  echo -e "${BL}Creating an Ubuntu 22.04 VM using the above default settings${CL}"
+  echo -e "${BL}Creating an Ubuntu VM using the above default settings${CL}"
 }
 
 function advanced_settings() {
+  if RELEASE=$(whiptail --backtitle "Proxmox VE Helper Scripts" --title "Ubuntu RELEASE" --radiolist "Choose Version" --cancel-button Exit-Script 10 58 3 \
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:169]:
```suggestion
  if RELEASE=$(whiptail --backtitle "Proxmox VE Helper Scripts" --title "UBUNTU VERSION" --radiolist "Choose Version" --cancel-button Exit-Script 10 58 2 \
```
consistent with LXC

File: vm/ubuntu-vm.sh [hunk 5]
```diff
@@ -155,10 +162,21 @@ function default_settings() {
...
+    "${ORACULAR}" "oracular  " OFF \
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:170]:
```suggestion
```

File: vm/ubuntu-vm.sh [hunk 6]
```diff
@@ -155,10 +162,21 @@ function default_settings() {
...
+    "${NOBLE}" "noble  " ON \
+    "${JAMMY}" "jammy  " OFF \
+    "${FOCAL}" "focal  " OFF \
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:173]:
```suggestion
```

File: vm/ubuntu-vm.sh [hunk 7]
```diff
@@ -155,10 +162,21 @@ function default_settings() {
...
   echo -e "${DGN}Start VM when completed: ${BGN}no${CL}"
-  echo -e "${BL}Creating an Ubuntu 22.04 VM using the above default settings${CL}"
+  echo -e "${BL}Creating an Ubuntu VM using the above default settings${CL}"
 }
 
 function advanced_settings() {
+  if RELEASE=$(whiptail --backtitle "Proxmox VE Helper Scripts" --title "Ubuntu RELEASE" --radiolist "Choose Version" --cancel-button Exit-Script 10 58 3 \
+    "${ORACULAR}" "oracular  " OFF \
+    "${NOBLE}" "noble  " ON \
+    "${JAMMY}" "jammy  " OFF \
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:172]: Move jammy option above noble, to be consistent with LXC

File: vm/ubuntu-vm.sh [hunk 8]
```diff
@@ -413,16 +431,16 @@ qm set $VMID \
...
-  msg_info "Starting Ubuntu 22.04 VM"
+  msg_info "Starting Ubuntu ${RELEASE} VM"
   qm start $VMID
-  msg_ok "Started Ubuntu 22.04 VM"
+  msg_ok "Started Ubuntu ${RELEASE} VM"
 fi
 msg_ok "Completed Successfully!\n"
 echo -e "Setup Cloud-Init before starting \n
-More info at https://github.com/tteck/Proxmox/discussions/2072 \n"
+More info at https://github.com/community-scripts/ProxmoxVE/discussions/268 \n"
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:446]:
```suggestion
More info at https://github.com/community-scripts/ProxmoxVE/discussions/272 \n"
```

File: vm/ubuntu-vm.sh [hunk 9]
```diff
@@ -130,11 +136,12 @@ function exit-script() {
 }
 
 function default_settings() {
+  RELEASE="${NOBLE}"
   VMID="$NEXTID"
   FORMAT=",efitype=4m"
   MACHINE=""
   DISK_CACHE=""
-  HN="ubuntu"
+  HN="ubuntu-${NOBLE}"
```
Comment (havardthom) [inline vm/ubuntu-vm.sh:144]: What happens to hostname if you choose jammy in advanced settings? you have to manually fix it?

**Comment by NoirPi** [comment]
Hello everyone,  

Thank you for the discussion and the proposed changes, especially by @havardthom. I would, however, like to suggest an alternative approach:  

Instead of having a single script for multiple versions, I’d prefer keeping separate scripts for each version. 
This makes it easier to target the desired version by simply adjusting the URL accordingly, which is at the moment just 2 digits to change:

- https://github.com/community-scripts/ProxmoxVE/raw/main/vm/ubuntu2204-vm.sh  
- https://github.com/community-scripts/ProxmoxVE/raw/main/vm/ubuntu2404-vm.sh

While I agree that the changes suggested by @havardthom make sense, they further reduce the need for a centralized script, especially since Ubuntu Focal will soon (Yes! 6 month is soon) reach its End of Standard Support and Oracular not being an LTS release.

This would again strip the script down to just 2 versions which is as shown above just a simple 2 digit change in the url.

For this reason, I would suggest not adding a unified script and instead sticking to separate scripts for each version.

**Comment by NoirPi** [comment]
> Should the Ubuntu LXC also be multiple scripts then? For me both of the approaches works, but we should pick one and be consistent. The approach we choose should also apply to all OS scripts in the future.

I would definitely suggest that we keep them separated.

The only thing that would be important now and in the future is the naming convention of the scripts themself for all OS Scripts this should also be consistent as you said.

**Comment by NoirPi** [comment]
I understand the argument, but I see it differently. Once I care about the version, I have to go into the advanced settings to select it. Up until now, this hasn’t been necessary.

Now, if we switch to a shared script and the version is only selected through advanced settings, the simple process of creating an Ubuntu 22.04 LXC would be changed.

What could be considered in my opinion is building the consolidated script in such a way that you can directly pass an argument for the version when calling it.

The option with the advanced settings is surely not that bad. But forcing into Advanced Settings to choose the Version, i dont think should be the way to go!

**Comment by havardthom** [comment]
Building a consolidated script would break the established conventions and patterns. Letting the user choose OS version without having to use advanced settings (and without having separate scripts for each version) is something we can consider in the future. As it stands today, all LXCs have OS version choice in advanced settings (like it has always been) and I think VMs should be the same until we choose to make that improvement.

**Comment by Mellowlynx** [comment]
> Building a consolidated script would break the established conventions and patterns. Letting the user choose OS version without having to use advanced settings (and without having separate scripts for each version) is something we can consider in the future. As it stands today, all LXCs have OS version choice in advanced settings (like it has always been) and I think VMs should be the same until we choose to make that improvement.

With a LXC, sure, but not with a VM, that should be set.
Also, when not adhering to the default OS, script support is discontinued.
So you can change it while deploying a LXC, but it has its price.

**Comment by havardthom** [comment]
Closing this since this is not something that needs to be fixed and there is no activity from author

**Pull request closed without merging**